src/lib/storage/reference_column/reference_column_iterable.hpp

```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
 
   template <typename Functor>
   void _on_with_iterators(const Functor& functor) const {
-    const auto table = _column.referenced_table();
-    const auto column_id = _column.referenced_column_id();
+    const auto pos_list = _column.pos_list();
+
+    const auto chunk_id_and_chunk_offsets_list = to_chunk_id_and_chunk_offsets_list(*pos_list);
+    if (chunk_id_and_chunk_offsets_list) {
+      const auto& [referenced_chunk_id, chunk_offsets_list] = *chunk_id_and_chunk_offsets_list;
+
+      _referenced_with_iterators(referenced_chunk_id, chunk_offsets_list, functor);
+      return;
+    }
 
-    const auto begin_it = _column.pos_list()->begin();
-    const auto end_it = _column.pos_list()->end();
+    const auto column_id = _column.referenced_column_id();
+    const auto table = _column.referenced_table();
 
-    auto begin = Iterator{table, column_id, begin_it, begin_it};
-    auto end = Iterator{table, column_id, begin_it, end_it};
+    auto begin = Iterator{column_id, table.get(), pos_list.get(), ChunkOffset{0u}};
+    auto end = Iterator{static_cast<ChunkOffset>(pos_list->size())};
     functor(begin, end);
   }
 
  private:
   const ReferenceColumn& _column;
 
  private:
+  template <typename Functor>
+  void _referenced_with_iterators(const ChunkID chunk_id, const ChunkOffsetsList& chunk_offsets_list,
+                                  const Functor& functor) const {
+    const auto column_id = _column.referenced_column_id();
```

Comment:
when writing code in the context of ReferenceColumns, I'd prefer it if the prefixes "referenced" and "referencing"/"reference" would be used in variable names. But that's just opinion, feel free to ignore